include/fuse_lowlevel.h

```diff
@@ -1160,6 +1160,34 @@ struct fuse_lowlevel_ops {
 	 */
 	void (*readdirplus) (fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
 			 struct fuse_file_info *fi);
+
+	/**
+	 * Copy a range of data from one file to another
+	 *
+	 * Performs an optimized copy between two file descriptors without the
+	 * additional cost of transferring data through the FUSE kernel module
+	 * to user space (glibc) and then back into the FUSE filesystem again.
+	 *
+	 * If this request is answered with an error code of ENOSYS, this is
+	 * treated as a permanent failure with error code EOPNOTSUPP, i.e. all
+	 * future copy_file_range() requests will fail with EOPNOTSUPP without
+	 * being send to the filesystem process.
+	 *
+	 * Valid replies:
+	 *   fuse_reply_copy_file_range
+	 *   fuse_reply_err
+	 *
+	 * @param req request handle
+	 * @param fd_in file description used for reading the data from
+	 * @param off_in starting point of fd_in from were the data is read
+	 * @param fd_out file description used for writing the data to
+	 * @param off_out starting point of fd_out where the data is written
+	 * @param len maximum size of the data to copy
+	 * @param flags passed along with the copy_file_range() syscall
+	 */
+	void (*copy_file_range) (fuse_req_t req, int fd_in, off_t off_in,
+				 int fd_out, off_t off_out, size_t len,
+				 int flags);
```

Comment:
This definition is incorrect – FUSE operations which operate on file descriptors take `fuse_file_info* fi` arguments, not `int fd`.

Additionally, all other FUSE low-level operations which operate on file descriptors are additionally given inode numbers (see, e.g, `fallocate` or `read`). The inode number is needed by the high-level interface to provide file path to the high-level callback. As such, in my opinion, this low-level callback should be actually declared as:
```c
void (*copy_file_range) (fuse_req_t req,
      fuse_ino_t ino_in, off_t off_in, fuse_file_info *fi_in,
      fuse_ino_t ino_out, off_t off_out, fuse_file_info *fi_out,
      size_t len, int flags);
```
There is a caveat though – the kernel module change proposed in https://lkml.org/lkml/2018/6/27/114 provides only `ino_in` argument because that is passed in the request's header (`args.in.h.nodeid = ff_in->nodeid;` in `fs/fuse/fuse.c` in the kernel) whereas `ino_out` is not given by the kernel to the userspace implementation. Isn't this an oversight in the kernel module? The lack of the inode number would make it impossible to create high-level interface which follows the libfuse's convention where the high-level callbacks are given paths to files which are to be modified.

Reply:
I'll see if I can add `ino_out` or change the file-descriptor to use `struct fuse_file_into*`.

The `ino_out` is not required for supporting the functionality with Gluster (does not use libfuse). If there is an advantage to having ino_out, it can be added to the kernel module. Do you want to send that suggestion as a reply to the kernel email so it is tracked there? I'll include your suggestion in v3 later this week.